api/hasura/actions/_handlers/vouch.ts

```diff
@@ -18,25 +19,56 @@ import {
 import { Awaited } from '../../../../api-lib/ts4.5shim';
 import { verifyHasuraRequestMiddleware } from '../../../../api-lib/validate';
 import {
+  composeHasuraActionRequestBodyWithApiPermissions,
   composeHasuraActionRequestBodyWithSession,
   HasuraUserSessionVariables,
+  vouchApiInput,
   vouchInput,
 } from '../../../../src/lib/zod';
 
 type Voucher = Awaited<ReturnType<typeof getUserFromProfileId>>;
 type Nominee = Awaited<ReturnType<typeof getNominee>>;
 
+const userRequestSchema = composeHasuraActionRequestBodyWithSession(
+  vouchInput,
+  HasuraUserSessionVariables
+);
+
+const apiRequestSchema = composeHasuraActionRequestBodyWithApiPermissions(
+  vouchApiInput,
+  ['create_vouches']
+);
+
+const requestSchema = z.union([userRequestSchema, apiRequestSchema]);
+
 async function handler(req: VercelRequest, res: VercelResponse) {
   try {
     const {
       input: { payload: input },
       session_variables: sessionVariables,
-    } = composeHasuraActionRequestBodyWithSession(
-      vouchInput,
-      HasuraUserSessionVariables
-    ).parse(req.body);
+    } = await requestSchema.parse(req.body);
+
+    let voucherProfileId: number | null;
+    // Allow passing in voucher_user_id for api-users to vouch on behalf of a user
+    if (sessionVariables.hasuraRole === 'api-user') {
+      assert('voucher_user_id' in input, 'voucher_user_id not specified');
+      voucherProfileId = input.voucher_user_id;
+
+      // Check if voucher exists in the same circle as the API key
+      // TODO: this uses assert for error handling
```

Comment:
others in the group seemed fine w/ using assert, do you have concerns about it? In fact you are using it above. I don't have a strong opinion but coming from other languages it seemed a bit strange. maybe its very normal in node/ts

Reply:
This is just copy pasted code from somewhere else, didn't write that comment myself

Reply:
lol i think i wrote it tbh